### PR TITLE
Move the key derivation code from helpers.Passphrase to crypto.FlexiKey

### LINKED ISSUE #6449: move kdf code
`borg.helpers.passphrase` module has a `Passphrase.kdf` method.

- that's crypto code, so it should rather be below `borg.crypto` package
- all the other code there is just rather simple input / output / env var processing

@hexagonrecursion ^^^

(master branch only)

### PR DESCRIPTION
**Pull request by hexagonrecursion**
Closes #6449

- [x] pbkdf2
- [x] argon2

@ThomasWaldmann, is this what you had in mind? If so, I'll migrate argon2 the same way.

### CODE REVIEW

File: src/borg/archiver.py [hunk 1]
```diff
@@ -626,7 +626,7 @@ def chunkit(chunker_name, *args, **kwargs):
         print("KDFs (slow is GOOD, use argon2!) ===============================")
         count = 5
         for spec, func in [
-            ("pbkdf2", lambda: Passphrase('mypassphrase').kdf(b'salt'*8, PBKDF2_ITERATIONS, 32)),
+            ("pbkdf2", lambda: FlexiKey.pbkdf2(Passphrase('mypassphrase'), b'salt'*8, PBKDF2_ITERATIONS, 32)),
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:629]: As Passphrase is just a subclass of `str` with some additional functionality (not relevant here), you can just give 'mypassphrase' directly.

File: src/borg/testsuite/crypto.py [hunk 2]
```diff
@@ -294,9 +294,8 @@ def test_decrypt_key_file_argon2_aes256_ctr_hmac_sha256(monkeypatch):
 def test_decrypt_key_file_pbkdf2_sha256_aes256_ctr_hmac_sha256(monkeypatch):
     plain = b'hello'
     salt = b'salt'*4
-    monkeypatch.setenv('BORG_PASSPHRASE', "hello, pass phrase")
-    passphrase = Passphrase.new()
-    key = passphrase.kdf(salt, iterations=1, length=32)
+    passphrase = Passphrase("hello, pass phrase")
+    key = FlexiKey.pbkdf2(passphrase, salt, 1, 32)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/crypto.py:298]: same here: if you remove the "read it from the env var" aspect, one could give a str directly to pbkdf2.

**Review by ThomasWaldmann — COMMENTED** [review]
LGTM, 2 minor things.

File: src/borg/crypto/key.py [hunk 3]
```diff
@@ -447,15 +450,53 @@ def decrypt_key_file(self, data, passphrase):
...
+    @staticmethod
+    def argon2(
+        passphrase,
+        output_len_in_bytes: int,
+        salt: bytes,
+        time_cost,
+        memory_cost,
+        parallelism,
+        type: Literal['i', 'd', 'id']
+    ) -> bytes:
```
Comment (ThomasWaldmann) [inline src/borg/crypto/key.py:468]: is it intended that you only gave types for some of the arguments?

File: src/borg/testsuite/crypto.py [hunk 4]
```diff
@@ -294,9 +294,8 @@ def test_decrypt_key_file_argon2_aes256_ctr_hmac_sha256(monkeypatch):
 def test_decrypt_key_file_pbkdf2_sha256_aes256_ctr_hmac_sha256(monkeypatch):
     plain = b'hello'
     salt = b'salt'*4
-    monkeypatch.setenv('BORG_PASSPHRASE', "hello, pass phrase")
-    passphrase = Passphrase.new()
-    key = passphrase.kdf(salt, iterations=1, length=32)
+    passphrase = "hello, pass phrase"
+    key = FlexiKey.pbkdf2(passphrase, salt, 1, 32)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/crypto.py:298]: do the same change in lines 285+ for argon2?

File: src/borg/crypto/key.py (hunk 3 above)
Comment (hexagonrecursion) [reply src/borg/crypto/key.py:468]: Good catch!

**Review by ThomasWaldmann — APPROVED** [review]

**Comment by ThomasWaldmann** [comment]
Thanks!

**Pull request merged**